### ISSUE TITLE
GUI: async Task ownership convention for long-running work

### ISSUE THREAD

**Issue by zajca** [zajca/zremote#34]
## Problem

The GUI has a mix of patterns for owning long-running async work:

- Some subsystems spawn a task with `.detach()` and rely on a `CancellationToken` for teardown.
- Some hold a `JoinHandle` (from tokio, not GPUI) in a struct field.
- Some use `Arc<Mutex<Option<JoinHandle>>>` with manual `abort()` calls.
- Some use `cx.background_spawn` without storing the handle at all, assuming the work will finish before the entity drops.

This inconsistency causes real problems:

1. **Leaks on window close.** If a poller is detached and the entity holding its "please stop" flag drops, the task may still run for a while, or forever if the cancellation path is broken.
2. **Races on reconnect.** Restarting a WebSocket poll after a network drop sometimes leaves the old task running alongside the new one, producing duplicate events.
3. **Unclear ownership.** A new contributor cannot tell from the code which async tasks belong to which entity and how they are stopped.
4. **Cannot reason about shutdown.** There is no single answer to "when the window closes, are all background pollers stopped?"

## What we want

**A single convention, enforced in review**, that says:

### Rule — Every long-running async task lives in a `Task<()>` field owned by a GPUI entity

```rust
// conceptual
pub struct FooPanel {
    ...
    poller: Option<Task<()>>,
}

impl FooPanel {
    pub fn start_poller(&mut self, cx: &mut Context<Self>) {
        // replacing any previous poller cancels it — Task drops when reassigned
        self.poller = Some(cx.spawn(|this, mut cx| async move {
            loop {
                // tick...
                if this.update(&mut cx, |this, cx| this.tick(cx)).is_err() {
                    break;   // entity dropped, exit task
                }
                cx.background_executor().timer(Duration::from_millis(500)).await;
            }
        }));
    }

    pub fn stop_poller(&mut self) {
        self.poller.take();   // drops the task, cancelling it
    }
}
```

**Why this works:**

- GPUI's `Task` is cancel-on-drop.
- Storing the task in the entity ties its lifetime to the entity. Window close → entity drop → task drop → cancel. No dangling pollers.
- `this.update(&mut cx, ...)` returns `Err` when the entity is gone, giving every poll loop an automatic exit condition.
- `Option<Task<()>>` lets the entity express "not running" without wrapping in a Mutex.
- Replacing the task (`self.poller = Some(new)`) cancels the old one atomically as a natural side effect of the drop — no explicit abort required.

### Rule — No `.detach()` for work that outlives a single tick

`.detach()` is fine for fire-and-forget short work (opening a URL, showing a toast). It is **not** fine for polling loops, WebSocket listeners, timers, or anything that mutates GUI state over time.

### Rule — Use `cx.background_spawn` for CPU / I/O, `cx.spawn` for state-touching tasks

`cx.spawn` gives access to the entity via `this.update(...)`. `cx.background_spawn` runs off the main thread but cannot touch entities directly. Choose based on whether the result needs to land in GPUI state.

### Rule — Helpers for the common cases

Provide small ergonomic helpers so every site doesn't reinvent the loop structure:

- `entity.spawn_poller(interval, on_tick)` — starts an interval poller, stores the Task in a provided slot, cancels previous.
- `entity.spawn_ws_listener(url, on_message, on_close)` — runs a WebSocket read loop with auto-reconnect, cancel-on-drop.

These are thin wrappers but they make the intent obvious at the call site.

### Rule — Audit and fix existing call sites

Sweep through `crates/zremote-gui/src` and convert:

- Every `.detach()` that runs a loop → move into a `Task<()>` field.
- Every `tokio::spawn` / `JoinHandle` → same.
- Every `Arc<Mutex<JoinHandle>>` → same, drop the Arc and Mutex.
- Every CancellationToken used only to stop background pollers from the GUI side → replace with Task drop.

CancellationToken still makes sense for **cross-crate** cancellation (server → agent, agent → session) — it is only the GUI-local use that this issue targets.

## Flows

**Window close**:
1. User closes the GUI window.
2. Main view entity drops.
3. Every owned panel drops.
4. Every `Task<()>` field drops.
5. Every async loop's next `this.update` returns Err → loop exits cleanly.
6. No orphans, no zombie WebSockets.

**Reconnect after network drop**:
1. Event WebSocket closes unexpectedly.
2. Reconnect logic calls `self.ws_listener = Some(new_task)`.
3. Previous task drops, cancels, exits.
4. Only the new task is running. No duplicate event delivery.

**Debug a misbehaving poller**:
1. Developer suspects a phantom poller.
2. Greps for `cx.spawn` / `cx.background_spawn` that don't store the result.
3. Finding none is a guarantee — convention is enforceable.

## Acceptance criteria

- [ ] Coding convention documented in `CLAUDE.md` under "GPUI Notes".
- [ ] `rust-reviewer` agent prompt updated to flag `.detach()` on long-running work.
- [ ] Every existing long-running `.detach()` / `JoinHandle` / `Arc<Mutex<JoinHandle>>` converted to a `Task<()>` field owned by its entity.
- [ ] No `tokio::spawn` from inside a GPUI view — all GUI async uses `cx.spawn` / `cx.background_spawn`.
- [ ] Helper methods `spawn_poller` and `spawn_ws_listener` (or equivalent) exist and are used where multiple call sites repeat the same pattern.
- [ ] Manual test: close the window while an agent is streaming events → no background activity after close (check via `tracing` spans).
- [ ] Manual test: trigger 5 WebSocket reconnects in a row → exactly one active listener after each reconnect (check duplicate events).

## Out of scope

- Cancellation of cross-crate work (server / agent). This issue is GUI-local.
- Replacing tokio under the hood (GPUI's background executor is still tokio-compatible).
- Introducing structured concurrency primitives beyond what GPUI already provides.

## Current state

- Mixed patterns across `crates/zremote-gui/src` — some detached, some JoinHandle, some Task, some Arc<Mutex>.
- No convention documented anywhere.
- No reviewer instruction to flag the mismatches.

## Related

- Prerequisite for #30 (Logs panel) and #29 (Chat panel) — both will add new pollers/streams, and starting with the wrong pattern guarantees technical debt on day one.
- Related to #25 (mDNS discovery) — the mDNS browser will run continuously and must use this pattern.

**Comment by zajca** [comment]
Related GUI infrastructure work:

- #31 — batched async persistence (its write loop uses this exact pattern)
- #32 — centralized keyboard dispatch
- #33 — render method decomposition

Feature dependencies:

- #25 — mDNS discovery browser is a long-running listener
- #27 — agent detection pollers / bell watchers
- #28 — managed processes and scheduled tasks spawn per-process pollers
- #29 — AI chat streaming SSE listener
- #30 — live log WebSocket tail
